debezium-core/src/main/java/io/debezium/config/Configuration.java

```diff
@@ -2125,18 +2125,19 @@ default String getFallbackStringProperty(Field newProperty, Field oldProperty) {
      * @return the evaluated value
      */
     static String getFallbackStringProperty(Configuration config, Field newProperty, Field oldProperty) {
-        if (null != config.getString(oldProperty.name()) && null != config.getString(newProperty.name())) { // both are set
-            CONFIGURATION_LOGGER.warn("Provided configuration has deprecated property \"" + oldProperty.name()
-                    + "\" and new property \"" + newProperty.name() + "\" set. Using value from \"" + newProperty.name() + "\"!");
+        if (null != config.getString(oldProperty.name())) {
+            if (null != config.getString(newProperty.name())) { // both are set
+                CONFIGURATION_LOGGER.warn("Provided configuration has deprecated property \"" + oldProperty.name()
+                        + "\" and new property \"" + newProperty.name() + "\" set. Using value from \"" + newProperty.name() + "\"!");
+            }
+            else { // old value is set
```

Comment:
This else condition here is the same as putting the warning into the lambda below, isn't it (it only triggers the?

That's what I meant with that this code change is not really a change of the behavior.

I suggest using the old implementation and adding a null check for the result of `config.getString(oldProperty)`. The warning should only happen when the oldProperty was set! (and pls check if the `getString()` call returns the default value when not set instead of null. and don't show the message in that case as well).

Reply:
If we put this else part inside lambda below it would log warning even when **no options are defined**, which is unnecessary. 

I think we don't have any default values for these fields, isn't it?

Reply:
>  it would log warning even when no options are defined

That's why I said we add a null check for the return value of `config.getString(oldProperty)`, similar to what you do. (but also checking for default return values, there should be an option to provide NULL as default value for `getString` in case of it's unset, maybe?!).

> I think we don't have any default values for these fields, isn't it?
This is an implementation in Configuration.java which is used everywhere and can potentially be used for properties that have a default value. But there should be implementations of `getString()` that allow to provide a manual default value that can be set to null, so instead of the defined default value for the field it will return null.

Reply:
@rk3rn3r Is this right by what you meant by returning the default value. 
```
 return config.getString(
                newProperty,
                () -> {
                    if (null != config.getString(oldProperty.name())) {
                        CONFIGURATION_LOGGER.warn("Using configuration property \"" + oldProperty.name()
                                + "\" is deprecated and will be removed in future versions. Please use \"" + newProperty.name()
                                + "\" instead.");
                        return config.getString(oldProperty);
                    }
                    else {
                        return oldProperty.defaultValueAsString();
                    }
                });
```

Reply:
sorry for the late reply, I had too many meetings yesterday in the second half of the working day-
I thought we do something like:

```
    static String getFallbackStringProperty(Configuration config, Field newProperty, Field oldProperty) {
        if (null != config.getString(oldProperty.name()) && null != config.getString(newProperty.name())) { // both are set
            CONFIGURATION_LOGGER.warn("Provided configuration has deprecated property \"" + oldProperty.name()
                    + "\" and new property \"" + newProperty.name() + "\" set. Using value from \"" + newProperty.name() + "\"!");
        }
        return config.getString(
                newProperty,
                () -> {
                    String oldValue = config.getString(oldProperty);
                    if (oldValue != null && !oldValue.equals(oldProperty.defaultValueAsString())) {
                        CONFIGURATION_LOGGER.warn("Using configuration property \"" + oldProperty.name()
                                + "\" is deprecated and will be removed in future versions. Please use \"" + newProperty.name()
                                + "\" instead.");
                    }
                    return oldValue;
                });
    }
```

or either:

```
    static String getFallbackStringProperty(Configuration config, Field newProperty, Field oldProperty) {
        if (null != config.getString(oldProperty.name()) && null != config.getString(newProperty.name())) { // both are set
            CONFIGURATION_LOGGER.warn("Provided configuration has deprecated property \"" + oldProperty.name()
                    + "\" and new property \"" + newProperty.name() + "\" set. Using value from \"" + newProperty.name() + "\"!");
        }
        return config.getString(
                newProperty,
                () -> {
                    String oldValue = config.getString(oldProperty, () -> null);
                    if (oldValue != null) {
                        CONFIGURATION_LOGGER.warn("Using configuration property \"" + oldProperty.name()
                                + "\" is deprecated and will be removed in future versions. Please use \"" + newProperty.name()
                                + "\" instead.");
                    }
                    return oldValue;
                });
    }
```

Reply:
I would also go with the first. Adding it and thereafter we can discuss duplicate warning part (DBZ-3218) whenever you are free. Thanks, @rk3rn3r :)